engine/execution/provider/engine.go

```diff
@@ -371,7 +358,7 @@ func (e *Engine) ensureAuthorized(chunkID flow.Identifier, originID flow.Identif
 	}
 
 	if origin.Weight == 0 {
-		return nil, engine.NewInvalidInputErrorf("node %s has zero weight at the block (%s) corresponding to chunk data pack (%s)", originID, blockID.String(), chunkID.String())
+		return nil, engine.NewInvalidInputErrorf("node %s has zero weight requesting chunk data pack", originID)
```

Comment:
what happens if execution falls behind finalization? Could this prevent VNs from verifying the last few blocks in the epoch?

Reply:
I think there are a couple of corner cases here:

1. A VN is newly staked, and execution is ahead of finalization. The VN won't be able to request the chunk data pack for the first blocks of the epoch until the block is finalized. Worst case the first blocks are delayed slightly, more likely other nodes assigned to that chunk provide enough approvals for consensus.
2. A VN is unstaking, and execution is behind finalization. The VN won't be able to verify the last blocks in the epoch since it has weight 0 at the latest finalized height, even though the block it's requesting is from its epoch.

I think in both cases, this is fine as long as there isn't a large % of VNs joining/leaving at the same epoch.

Reply:
This was the main reason why we directly asked whether a VN is staked at the respective blocks: 
* it avoids all this ambiguity and edge cases that Peter is discussing
* Asking "latest finalized" is an expensive operation, because it must be atomic across the entire system. Meaning, we _always_ resort to a data base read. 

   In comparison, the answers to queries about specific blocks are cached in most cases. Likely, the blocks that VNs are requesting were recently executed. Recently executed blocks already reside in the caches, including what nodes are authorized to process this block. Hence, the runtime cost of querying this information is _very_ cheap. 
  
Note: the block-specific resolution of VNs is the most precise. Moving to "nodes at the most recently finalized block" is a coarser simplification, i.e. it adds edge cases and therefore complexity and surface area for bugs (much more stuff can go wrong).  
  
So the reasoning to switch from "block-specific" to "most recently finalized" is only worth it if it _significantly_ reduces implementation complexity otherwise.

Reply:
After more offline discussions with Alex, we have decided that we don't need this logic as its already provided and checked by the networking layer (https://github.com/onflow/flow-go/blob/master/network/channels/channels.go#L210) 

We might do extra levels of caching or a special type of validation of requests in the future, but that's outside of the scope of this PR.